packages/warning/src/index.js

```diff
@@ -4,11 +4,7 @@
 import { logged } from './utils';
 
 function isDev() {
-	return (
-		typeof process !== 'undefined' &&
-		process.env &&
-		process.env.NODE_ENV !== 'production'
-	);
```

Comment:
Even though the `process` polyfill is no longer present, it shouldn't matter for `process.env.NODE_ENV`, because it's a special constant defined and replaced by the `DefinePlugin`. This plugin will expand
```js
process.env.NODE_ENV !== 'production'
```
or
```js
process?.env?.NODE_ENV !== 'production'
```
into
```js
'development' !== 'production'
```
eliminating the `process` reference from the output.

But the `DefinePlugin` doesn't undestand what `process` or `process.env` is and will leave it in the output.

Therefore, the easiest fix for the bug is to use:
```js
return ( process?.env?.NODE_ENV ?? 'production' ) !== 'production';
```
Here `DefinePlugin` can do the expansion and keeps the logic where:
- if `process` or `process.env` doesn't exist, `isDev` is always `false`.
- otherwise `isDev` is `true` iff `NODE_ENV` is not `production`.

Adding support for `SCRIPT_DEBUG` is nice, but I think it's a mistake to assume that `process.env.NODE_ENV` is bad and no longer supported.

Reply:
> Therefore, the easiest fix for the bug is to use:
> 
> ```js
> return ( process?.env?.NODE_ENV ?? ‘production’ ) !== ‘production’;
> ```

Yes, but it only fixes the bug for the `@wordpress/warning` usage with the integrated Babel plugin. It doesn’t help with the case where someone wants to use `process.env.NODE_ENV` in other packages to guard the usage of code. They would have to use something like ` ( process?.env?.NODE_ENV ?? ‘production’ ) !== ‘production’` to avoid errors when the `DefinePlugin` is not replacing the entry. Therefore I think that `SCRIPT_DEBUG` is way simpler to use. Folks outside of WordPress core and the Gutenberg plugin can use the longer version with `DefinePlugin`, if they prefer it over `SCRIPT_DEBUG`.

Reply:
> to avoid errors when the `DefinePlugin` is not replacing the entry.

When would such an error happen? It won't happen in Node.js, and it won't happen in webpack (if `process.env.NODE_ENV` is written in such a way that `DefinePlugin` can fully replace it). It can happen in a browser, when the `@wordpress/warning` package is imported with the native ES module loader. Is that the case we try to guard against here?

It's increasingly common that packages ship separate ES modules for Node and for browsers. Specified in the `exports` field. If we decided to do that, too, presence of `process.env.NODE_ENV` would be on of the differences.

When testing this PR, I see one odd thing: the built script for the `warning` package, in `build/warning/index.js`, the `SCRIPT_DEBUG` variable is replaced away, and in production version the `warning` function is empty. `SCRIPT_DEBUG` global will never enable it again.

So, do we really want to specify `SCRIPT_DEBUG` with `DefinePlugin`? Wouldn't it be better to keep the references to the global in the built script?

Reply:
> the SCRIPT_DEBUG variable is replaced away

that is the expected behavior with this plugin, yes. It‘s a global that‘s compiled away during build

Reply:
> > to avoid errors when the DefinePlugin is not replacing the entry.
> When would such an error happen? It won't happen in Node.js, and it won't happen in webpack (if process.env.NODE_ENV is written in such a way that DefinePlugin can fully replace it).

It's the bug that was filed initially in https://github.com/WordPress/gutenberg/issues/44950. The current webpack configuration doesn't do anything with `process.env.NODE_ENV`. 

> When testing this PR, I see one odd thing: the built script for the warning package, in build/warning/index.js, the SCRIPT_DEBUG variable is replaced away, and in production version the warning function is empty. SCRIPT_DEBUG global will never enable it again.

Yes, that is expected as it aligns with how WordPress core loads files. See a more detailed explanation in https://github.com/WordPress/gutenberg/pull/50122#issuecomment-1551258038.